src/EFCore.PG/Storage/Internal/Mapping/NpgsqlHstoreTypeMapping.cs

```diff
@@ -56,19 +59,36 @@ protected override string GenerateNonNullSqlLiteral(object value)
             return sb.ToString();
         }
 
-        sealed class HstoreComparer : ValueComparer<Dictionary<string, string>>
+        static ValueComparer GetComparer(Type clrType)
+        {
+            if (clrType == typeof(Dictionary<string, string>))
+                return MutableComparerInstance;
+
+            if (clrType == typeof(ImmutableDictionary<string, string>))
+            {
+                // Because ImmutableDictionary is immutable, we can use the default value comparer, which doesn't
+                // clone for snapshot and just does reference comparison.
+                // We could compare contents here if the references are different, but that would penalize the 99% case
+                // where a different reference means different contents, which would only save a very rare database update.
+                return null;
+            }
+
+            throw new ArgumentException($"CLR type must be {nameof(Dictionary<string,string>)} or {nameof(ImmutableDictionary<string,string>)}");
+        }
+
+        sealed class HstoreMutableComparer : ValueComparer<Dictionary<string, string>>
         {
-            public HstoreComparer() : base(
+            public HstoreMutableComparer() : base(
                 (a, b) => Compare(a,b),
                 o => o.GetHashCode(),
                 o => o == null ? null : new Dictionary<string, string>(o))
             {}
 
-            static bool Compare(Dictionary<string, string> a, Dictionary<string, string> b)
+            static bool Compare(IReadOnlyDictionary<string, string> a, IReadOnlyDictionary<string, string> b)
```

Comment:
You have `IReadOnlyDictionary<string, string>` here, but the comparer is for `Dictionary<string, string>`. Therefore, you're loosing a chance to inline some calls.

Reply:
Done, thanks.